simplify handling of replication failure notification

welansari opened the pull request:
Instead of duplicating the notification logic we set the correct originOp when updating the metadata to trigger the replication failure notification. (Bucket notification listens to the oplog and sends notification based on the originOp field in the metadata)

At the end of a failed replication we update the metadata with the failure status that ends up in the oplog topic. So this does not add any overhead in terms of extra Kafka messages being pushed to any of the topics.

This also fixes the duplicate notifications bug where we send as many notifications as destinations we have in case they use the same internal topic (bug already fixed in notification code).

Issue: BB-507

Commit:
simplify handling of replication failure notification

Instead of duplicating the notification logic we set the
correct originOp when updating the metadata to trigger the
replication failure notification.

At the end of a failed replication we update the metadata
with the failure status that ends up in the oplog topic.
So this does not add any overhead in terms of extra Kafka messages
being pushed to any of the topics.

This also fixes the duplicate notifications bug where we send as many
notifications as destinations we have in case they use the same internal
topic.

Issue: BB-507

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-507 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.39`

* `8.7.0`


Please check the `Fix Version/s` of BB-507, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

francoisferrand commented on lib/models/ObjectQueueEntry.js:
```diff
@@ -193,21 +194,24 @@ class ObjectQueueEntry extends ObjectMD {
             .setReplicationSiteStatus(site, 'COMPLETED')
-            .setReplicationStatus(this._getGlobalReplicationStatus());
+            .setReplicationStatus(this._getGlobalReplicationStatus())
+            .setOriginOp('');
```
should we not set a value anyway?

it could (eventually) be used to generate non-standard notification (or make it easier/safer to refactor), and will be helpful already for troubleshooting: as we can directly see what caused the metadata change...

(note: need to make sure it does not match one of the "wildcard" notification rules, though...)

francoisferrand commented on lib/models/ObjectQueueEntry.js:
```diff
@@ -193,21 +194,24 @@ class ObjectQueueEntry extends ObjectMD {
             .setReplicationSiteStatus(site, 'PENDING')
-            .setReplicationStatus(this._getGlobalReplicationStatus());
+            .setReplicationStatus(this._getGlobalReplicationStatus())
+            .setOriginOp('');
```
same

Commit:
filter out non supported notification

Notification rules are normally checked when setting the notification
configuration (even wildcards), however we need an explicit check at
the queue populator level to filter out non standard or unsupported
notification events that might validate one or more of the wildcard
rules set.

Issue: BB-507

francoisferrand commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -91,6 +91,18 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
+     */
+    _isNotificationEventSupported(eventType) {
+        return supportedNotificationEvents.has(eventType);
+    }
```
not really an issue, but we may (at some point, just maybe) want to allow explicitely registering for these non-standard notifications (even though we probably don't want those to be picked up automatically by wildcard configs indeed)
→ we should remember this, may trick us later... (but certainly fine for now)

francoisferrand commented on tests/unit/replication/UpdateReplicationStatus.js:
```diff
@@ -88,6 +86,20 @@ describe('UpdateReplicationStatus._getUpdatedSourceEntry()', () => {
         assert.strictEqual(updatedSourceEntry, null);
     });
+
+    it('should return a PENDING entry when site status is FAILED', () => {
```
.... `and replay count not reached` ?

(with this title it looks like a bug, and it seems we also have the case where it returns FAILED...)

welansari replied:
Removed the test as it's a case that can never happen and can be confusing

Commit:
bump version to 8.6.39

Issue: BB-507

Kerkesni commented:
/approve

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-507` into
`development/8.6`
* [w/8.7/bugfix/BB-507](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/bugfix/BB-507) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

Kerkesni commented:
/wait

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

welansari commented:
/approve

bert-e commented:
# History mismatch



Merge commit #7bf30e15e1c854a162e1feeb593d4fce0841588c on the integration branch
`w/8.7/bugfix/BB-507` is merging a branch which is neither the current
branch `bugfix/BB-507` nor the development branch
`development/8.7`.

It is likely due to a rebase of the branch `bugfix/BB-507` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

welansari commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-507` into
`development/8.6`
* [w/8.7/bugfix/BB-507](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/bugfix/BB-507) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



Please check the status of the associated issue BB-507.

Goodbye kerkesni.




*The following options are set:* **approve**
